common/processors/ingest_local_file.py

```diff
@@ -0,0 +1,265 @@
+        options = super(CSVIngestor, self)._options
+        options.update(encoding=self._encoding)
+        return options
+
+    @property
+    def _post_processors(self):
+        return [self._skip_header, self._drop_bad_rows]
+
+    @cached_property
+    def _encoding(self):
+        """Select or detect the file encoding and set the resource utf-8."""
+
+        if self.resource.get('encoding'):
+            return self.resource['encoding']
+        else:
+            return self._detect_encoding()
+
+    def _detect_encoding(self):
+        """Sniff the encoding using the entire file."""
+
+        with open(self.resource['path'], 'rb') as stream:
+            text = stream.read()
+            encoding = chardet.detect(text)['encoding']
+            info('Detected %s encoding with chardet', encoding)
+            return encoding
+
+    @property
+    def _raw_headers(self):
+        """Return headers as found in the data file."""
+
+        options = dict(headers=1, encoding=self._encoding)
+        with Stream(self.resource['path'], **options) as stream:
+            return stream.headers
+
+    @staticmethod
+    def _drop_bad_rows(rows):
+        """Drop rows when they don't match headers (post-processor)."""
+
+        for index, headers, row in rows:
+            if len(row) == len(headers):
+                yield index, headers, row
+            else:
+                warning('Bad row %s = %s', index, format_to_json(row))
+
+    @staticmethod
+    def _skip_header(rows):
+        """Skip the header (post-processor)."""
+
+        # Headers are passed as an option and need to be explicitly ignored.
+        for index, headers, row in rows:
+            if index != 1:
+                yield index, headers, row
+
+
+class JSONIngestor(BaseIngestor):
+    """An ingestor for json files."""
+
+    @property
+    def _pre_processors(self):
+        return [self._fill_missing_fields]
+
+    @property
+    def _fill_missing_fields(self):
+        """Pre-fill incomplete JSON rows (to avoid fields mixing up)."""
+
+        def fill_missing_fields():
+            with open(self.resource['path']) as stream:
+                rows = json.loads(stream.read())
+
+            for row in rows:
+                for header in self._headers:
+                    if header not in row:
+                        row[header] = None
+
+            with open(self.resource['path'], 'w+') as stream:
+                stream.write(format_to_json(rows))
+
+        return fill_missing_fields
+
+    @property
+    def _raw_headers(self):
+        """Return all field names encountered in the file."""
+
+        keys = set()
+
+        with open(self.resource['path']) as stream:
+            rows = json.loads(stream.read())
+            for row in rows:
+                keys.update(set(row))
+
+        return sorted(keys)
+
+
+class XLSIngestor(BaseIngestor):
+    """An ingestor for xls and xlsx files."""
+
+    @staticmethod
+    def force_strings(rows):
```

Comment:
I'm writing them as i need them. So far, I haven't had a use case. As I said, I'm not outright replacing the old module. I'm introducing it gradually.